### PR TITLE
Creation of the ros wrapper 

### PR DESCRIPTION
**Pull request by miguelvidex**
Based on the demo "video_emotion_gender_demo.py" this is a ROS node to use this application for face classification.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	message_generation
+)
+
+catkin_python_setup()
+
+## Generate messages in the 'msg' folder
+add_message_files(
+	DIRECTORY ros/msg
+	FILES
+	face_classification_single.msg
```
Comment (socrob) [inline CMakeLists.txt]: Please follow the [ROS naming conventions](http://wiki.ros.org/ROS/Patterns/Conventions#Messages).

FaceClassification + FaceClassificationArray

File: package.xml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0"?>
+<package>
+ <name>face_classification</name>
+ <version>1.0.0</version>
+ <description>
+ TODO!
```
Comment (socrob) [inline package.xml]: Add description

File: ros/doc/README.md [hunk 3]
```diff
@@ -0,0 +1 @@
+mbot_face_classification documentation
```
Comment (socrob) [inline ros/doc/README.md]: Remove file

File: ros/config/config_face_classification.yaml [hunk 4]
```diff
@@ -0,0 +1 @@
+# TODO!!
```
Comment (socrob) [inline ros/config/config_face_classification.yaml]: Remove file if unused

File: package.xml [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+ <license>GPLv3</license>
+
+ <author email="miguel.oliveira.silva@tecnico.ulisboa.pt">Miguel Silva</author>
+ <maintainer email="miguel.oliveira.silva@tecnico.ulisboa.pt">Miguel Silva</maintainer>
+
+ <buildtool_depend>catkin</buildtool_depend>
+
+ <build_depend>rospy</build_depend>
+ <build_depend>message_generation</build_depend>
+
```
Comment (socrob) [inline package.xml]: Please add dependencies here and CMakeLists or make a note to add in the near future.

File: CMakeLists.txt (hunk 1 above)
Comment (socrob) [inline CMakeLists.txt]: Please adhere to the [ROS naming conventions](http://wiki.ros.org/ROS/Patterns/Conventions#Messages) e.g. FaceClassification and FaceClassificationArray

File: CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+	face_classification_array.msg
+)
+
+add_service_files(
+  DIRECTORY ros/srv
+  FILES 
+  face_classification_buffered.srv
```
Comment (socrob) [inline CMakeLists.txt]: Also naming conventions here

File: package.xml (hunk 2 above)
Comment (socrob) [inline package.xml]: Add description

File: package.xml (hunk 5 above)
Comment (socrob) [inline package.xml]: Add dependencies + in CMakeLists

File: ros/config/config_face_classification.yaml (hunk 4 above)
Comment (socrob) [inline ros/config/config_face_classification.yaml]: Remove if unused

File: ros/doc/README.md (hunk 3 above)
Comment (socrob) [inline ros/doc/README.md]: Remove file

File: package.xml (hunk 2 above)
Comment (socrob) [inline package.xml]: Add a small README file, outside of ros directory. With dependencies and quick installation instructions and how to run.

**Review by ghost — CHANGES_REQUESTED** [review]

File: package.xml (hunk 2 above)
Comment (socrob) [reply package.xml]: Unrelated to this file but the comment is relevant.

**Comment by miguelvidex** [comment]
Can you give me some feedback about node structure? 
https://github.com/miguelvidex/face_classification/blob/master/ros/src/face_classification_ros/face_classification_node.py

And I already did some changes related to what you commented, take a look if possible.

File: CMakeLists.txt [hunk 7]
```diff
@@ -3,19 +3,18 @@ project(face_classification)
 
 find_package(catkin REQUIRED COMPONENTS
 	rospy
-  	std_msgs
+  std_msgs
```
Comment (socrob) [inline CMakeLists.txt]: Fix indentation, also line 9

File: CMakeLists.txt [hunk 8]
```diff
@@ -25,6 +24,8 @@ add_service_files(
   face_classification_continuous.srv
```
Comment (socrob) [inline CMakeLists.txt]: Make the same changes for .srv that you did for .msg, as discussed.

File: ros/src/face_classification_ros/face_classification_node.py [hunk 9]
```diff
@@ -0,0 +1,401 @@
...
+    # Rate of the image acquisition in buffer mode
+    self.image_acquisition_rate = rospy.Rate(rospy.get_param("image_acquisition_rate",4))
+
+    # hyper-parameters for bounding boxes shape
+    self.frame_window = rospy.get_param("~frame_window",10)
+    self.gender_offsets = rospy.get_param("~gender_offsets",(30, 60))
+    self.emotion_offsets = rospy.get_param("~emotion_offsets",(20, 40))
+
+    # get an instance of RosPack with the default search paths
+    self.rospack = rospkg.RosPack()
```
Comment (guilhermelawless) [inline ros/src/face_classification_ros/face_classification_node.py]: Doesn't need to be `self`

File: ros/src/face_classification_ros/face_classification_node.py [hunk 10]
```diff
@@ -0,0 +1,401 @@
...
+        faces_genders - list of gender for each face
+        faces_emotions - list of emotions for each face
+        image_format - the encoding of the image data
+    '''
+    #time_stamp of published msgs
+    time_stamp = rospy.Time.now()
+
+    # Array of detected faces
+    faces_array = FaceClassificationArray()
+    faces_array.header=Header()
```
Comment (guilhermelawless) [inline ros/src/face_classification_ros/face_classification_node.py]: No need for this, above line already creates the Header object.

File: ros/src/face_classification_ros/face_classification_node.py [hunk 11]
```diff
@@ -0,0 +1,401 @@
...
+    faces_array.header.stamp = time_stamp
+
+    # for each face create the msg and add it to the face_array
+    for idx_face, face_coordinates in enumerate(faces_coordinates):
+      new_face = FaceClassification()
+      # face id equal to the number of face detected by default
+      new_face.id=idx_face+1
```
Comment (guilhermelawless) [inline ros/src/face_classification_ros/face_classification_node.py]: I think we should use 0-indexed as seen in object detection wrappers. Also applies to line 208.

**Review by guilhermelawless — CHANGES_REQUESTED** [review]
Full review of the node.

**Comment by guilhermelawless** [comment]
@miguelvidex  i have been thinking more of the best solution to sync the output msg to the input image. Let's imagine detection takes around 5 seconds (not the case of course).

Right now the output msg has a header with `rospy.Time.now()` stamp. I believe this is correct. However, how can the subscribing node know to which image this detection corresponds? I think my previous idea of using the image stamp is bad because in cases of ROS Bag replays, it will be there before it should (5s delay won't be there).

I think the best approach is the one taken here: https://github.com/leggedrobotics/darknet_ros/blob/master/darknet_ros_msgs/msg/BoundingBoxes.msg , i.e., use another field with the corresponding input image header. This should be only for the Array msg.

Do you agree with this approach?

**Pull request merged**